nnsmith/materialize/torch/parse.py

```diff
@@ -0,0 +1,171 @@
+
+    named_modules = dict(gm.named_modules())
+    ir = GraphIR()
+    name_2_retvals: Dict[str, List[str]] = {}
+    for i_node, node in enumerate(gm.graph.nodes):
+        node = cast(fx.node.Node, node)
+        if node.op == "placeholder":
+            iexpr = InstExpr(Input(dim=len(node.meta["res"].shape)), [])
+        else:
+            args_flatten, args_treespec = pytree.tree_flatten(node.args)
+            kwargs_flatten, kwargs_treespec = pytree.tree_flatten(node.kwargs)
+            input_nodes = [
+                a
+                for a in (args_flatten + kwargs_flatten)
+                if isinstance(a, fx.node.Node)
+            ]
+            input_valstrs = list(map(lambda n: name_2_retvals[n.name][0], input_nodes))
+            input_like = list(
+                map(
+                    lambda ts: AbsTensor(
+                        shape=ts.shape, dtype=DType.from_torch(ts.dtype)
+                    ),
+                    pytree.tree_flatten(
+                        list(map(lambda n: n.meta["res"], input_nodes))
+                    )[0],
+                )
+            )
+            output_like = list(
+                map(
+                    lambda ts: AbsTensor(
+                        shape=ts.shape, dtype=DType.from_torch(ts.dtype)
+                    ),
+                    pytree.tree_flatten(node.meta["res"])[0],
+                )
+            )
+            args_wo_nodes = pytree.tree_map(
+                lambda n: ConcreteOp.empty if isinstance(n, fx.node.Node) else n,
+                node.args,
+            )
+            kwargs_wo_nodes = pytree.tree_map(
+                lambda n: ConcreteOp.empty if isinstance(n, fx.node.Node) else n,
+                node.kwargs,
+            )
+            if node.op == "call_function":
+                if (
+                    node.target is operator.getitem
+                    and isinstance(node.args[0], fx.node.Node)
+                    and not isinstance(node.args[0].meta["res"], torch.Tensor)
+                ):
+                    name_2_retvals[node.name] = [
+                        name_2_retvals[node.args[0].name][node.args[1]]
+                    ]
+                    continue
+                else:
+                    target_str = node._pretty_print_target(node.target)
+            elif node.op == "call_method":
+                target_str = f"torch.Tensor.{node.target}"
+            elif node.op == "call_module":
+                target = named_modules[node.target]
+                target_str = repr(target)
+                if target.__module__.startswith("torch.nn.modules"):
+                    target_str = f"torch.nn.{target_str}"
+            elif node.op == "get_attr":
+                raise NotImplementedError(f"{node.op = }, {node.name = }")
+            elif node.op == "output":
+                continue
+            else:
+                raise ValueError(f"Unexpected {node.op = }")
```

Comment:
Same here.

Reply:
Please address `{node.op = }`. We can do another PR to upgrade the codebase to 3.8 (as TF and PT are deprecating 3.7). But let's move a bit slower behind them until no one else is interested in testing old code.